### ISSUE TITLE
OO Event + plugin hook handlers

### ISSUE THREAD

**Issue by ewinslow** [Elgg/Elgg#7461]
I'm thinking we do something like:

``` php
interface EventHandler {
  public function handleEvent($event, $type, $object);
}

interface HookHandler {
  public function handleHook($hook, $type, $result, $params);
}
```

Eventually I'd want it to work like:

``` php
elgg_register_event_handler('event', 'type', EventHandlerImpl::class);
elgg_register_plugin_hook_handler('event', 'type', HookHandlerImpl::class);
```

Then they can be dependency injected and just call the `handle*` method when appropriate

In any case, we can still convert the logic of our current handlers to Handler-like classes, even if they don't technically implement the same interface or whatever.
## Rant on the current plugin hook/event api:

My feeling personally is that eventually we'll want something more specific than plugin hooks for everything. They are a good general solution if you just want quick extensibility, but I guess my opinion lately has been to prefer more custom-tailored APIs for each case where you want extensibility.

Type saftey has become more attractive in my mind over the past couple years because it's an easy way to avoid having to do so much defensive coding, and avoid writing repetetive/worthless tests to check that your defensive code works.

Plugin hooks definitely aren't built for type safety or self-describing APIs. I feel like I have to memorize `$params` for every plugin hook manually or else look up the documentation instead of just being able to tell based on a type-hint.

**Comment by juho-jaakkola** [comment]
The Container Aware Event Dispatcher was removed in Symfony 4. The [EventDispatcher](http://symfony.com/doc/current/components/event_dispatcher.html) component should be used instead. I have some pretty good experience of it being used in Drupal 8.

**Comment by jdalsem** [comment]
Event and hook handlers need to be callables. You can use a `MyClass::class` callback as demonstrated here https://github.com/Elgg/Elgg/blob/master/mod/likes/start.php#L36

**Comment by hypeJunction** [comment]
You can get elgg() from the event/hook object, so it's a merger of finishing the DI PR

**Comment by jdalsem** [comment]
Closed in favor of #14117
